internal/service/dynamodb/global_secondary_index.go

```diff
@@ -0,0 +1,912 @@
+				err.Error(),
+			)
+		}
+
+		return
+	}
+
+	ut := &dynamodb.UpdateTableInput{
+		TableName: data.Table.ValueStringPointer(),
+		GlobalSecondaryIndexUpdates: []awstypes.GlobalSecondaryIndexUpdate{
+			{
+				Delete: &awstypes.DeleteGlobalSecondaryIndexAction{
+					IndexName: data.IndexName.ValueStringPointer(),
+				},
+			},
+		},
+	}
+
+	if _, err = conn.UpdateTable(ctx, ut); err != nil {
+		response.Diagnostics.AddError(
+			fmt.Sprintf(`Unable to delete index "%s" on table "%s"`, data.IndexName.ValueString(), data.Table.ValueString()),
+			err.Error(),
+		)
+
+		return
+	}
+
+	if _, err := waitGSIDeleted(ctx, conn, data.Table.ValueString(), data.IndexName.ValueString(), deleteTimeout); err != nil {
+		response.Diagnostics.AddError(
+			fmt.Sprintf(`Error while waiting for GSI "%s" on table "%s" to be deleted`, data.IndexName.ValueString(), data.Table.ValueString()),
+			err.Error(),
+		)
+	}
+}
+
+type resourceGlobalSecondaryIndexModel struct {
+	framework.WithRegionModel
+
+	ARN                 types.String                                             `tfsdk:"arn"`
+	HashKey             types.String                                             `tfsdk:"hash_key"`
+	HashKeyType         fwtypes.StringEnum[awstypes.ScalarAttributeType]         `tfsdk:"hash_key_type"`
+	IndexName           types.String                                             `tfsdk:"index_name"`
+	NonKeyAttributes    fwtypes.SetOfString                                      `tfsdk:"non_key_attributes"`
+	ProjectionType      fwtypes.StringEnum[awstypes.ProjectionType]              `tfsdk:"projection_type"`
+	RangeKey            types.String                                             `tfsdk:"range_key"`
+	RangeKeyType        fwtypes.StringEnum[awstypes.ScalarAttributeType]         `tfsdk:"range_key_type"`
+	ReadCapacity        types.Int64                                              `tfsdk:"read_capacity"`
+	Table               types.String                                             `tfsdk:"table"`
+	WriteCapacity       types.Int64                                              `tfsdk:"write_capacity"`
+	OnDemandThroughputs fwtypes.ListNestedObjectValueOf[onDemandThroughputModel] `tfsdk:"on_demand_throughput"`
+	WarmThroughputs     fwtypes.ListNestedObjectValueOf[warmThroughputModel]     `tfsdk:"warm_throughput"`
+
+	Timeouts timeouts.Value `tfsdk:"timeouts"`
+}
+
+func (data *resourceGlobalSecondaryIndexModel) Flatten(ctx context.Context, v any) diag.Diagnostics {
+	var diags diag.Diagnostics
+
+	g, ok := v.(awstypes.GlobalSecondaryIndexDescription)
+	if !ok {
+		diags.AddError(
+			"Bad argument",
+			"Given parameter is not awstypes.GlobalSecondaryIndexDescription",
+		)
+
+		return diags
+	}
+
+	table, ok := ctx.Value(tableKey).(*awstypes.TableDescription)
+	if !ok || table == nil {
+		diags.AddError(
+			"Bad argument",
+			"Context is missing the table description",
+		)
+
+		return diags
+	}
+
+	data.ARN = types.StringValue(aws.ToString(g.IndexArn))
```

Comment:
Replace `types.StringValue(aws.ToString(val))` with `fwflex.StringToFramework(val)` to get proper handling including `nil` values

```suggestion
	data.ARN = fwflex.StringToFramework(g.IndexArn)
```